### PR TITLE
Tl/parser prototyp

### PR DESCRIPTION
**Pull request by Torpheus**
Erster Prototyp des Schemeparsers

### CODE REVIEW

File: src/Parser.hs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+data MetaNode = LambdaNode [MetaNode] MetaNode [MetaNode] | PairNode MetaNode MetaNode 
+            | RealAtom Double | IntegralAtom Int | RationalAtom Int Int | EmptyAtom
+            | StringAtom String | ComplexAtom (Complex Double) | BoolAtom Bool | CharAtom Char 
+            | IdentifierAtom String | ApplicationNode MetaNode [MetaNode] 
+            | IfNode MetaNode MetaNode MetaNode | SetNode MetaNode MetaNode | DefineNode MetaNode MetaNode 
+            deriving (Eq, Show)
+
+parseScheme :: [Token] -> MetaNode
+parseScheme st = case runState (parseExpression "Scheme Program")  st of
```
Comment (rapgru) [inline src/Parser.hs]: Unnecessary space

File: src/Parser.hs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    (mn, []) -> mn
+    (_, t:_) -> error $ "Unexpected token " ++ show t ++  " not allowed in top level of program"
```
Comment (rapgru) [inline src/Parser.hs]: Unnecessary space

File: src/Parser.hs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    put (t:ts)
+
+pull :: String -> State [Token] Token
+pull context = do
+    ts <- get
+    case ts of 
+        (t:ts) -> do
+            put ts
+            return t
+        _ -> error $ "Unexpected end of token stream in " ++ context
```
Comment (rapgru) [inline src/Parser.hs:43]: Error call with exactly this message is duplicated in pullEq and peek

File: src/Parser.hs [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+pullEq :: String -> Token -> State [Token] ()
+pullEq context c = do
```
Comment (rapgru) [inline src/Parser.hs:46]: pullEq is a variant of pull - maybe you can reuse pull inside of pullEq

File: src/Parser.hs [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    pullEq "Syntactic construct" POpen 
+    t <- peek "Syntactic construct"
+    case t of
+        Lambda -> parseLambda
+        Define -> parseDefine
+        Quote -> parseQuote
+        QuasiQuote -> parseQuasiQuote
+        If -> parseIf
+        Set -> parseSet
+        _ -> parseApplication
```
Comment (rapgru) [inline src/Parser.hs:89]: can this be `Identifier -> parseApplication` followed by `_ -> error ...`?

File: src/Parser.hs [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+            t <- pull "Formal Paramterlist"
+            case t of 
+                Identifier str -> do
+                    pullEq "Formal Paramterlist" PClose
+                    return ([], IdentifierAtom str)
+                _ -> error $ "Expected formal parameters in formal parameterlist not token " ++ show t
+        Identifier str -> do
+            (is, i) <- parseFormalParameterList
+            return ((IdentifierAtom str):is, i)
+        _ -> error $ "Expected formal parameters in formal parameterlist not token " ++ show t
```
Comment (rapgru) [inline src/Parser.hs]: add a newline

**Review by rapgru — CHANGES_REQUESTED** [review]
Small details to be accounted...

File: src/Parser.hs (hunk 5 above)
Comment (Torpheus) [reply src/Parser.hs:89]: No the Scheme grammar does not allow this

File: src/Parser.hs (hunk 4 above)
Comment (Torpheus) [reply src/Parser.hs:46]: Good idea

**Review by rapgru — APPROVED** [review]

**Pull request merged**